### PR TITLE
[DataGrid] Pass `rowId` param to `processRowUpdate`

### PR DESCRIPTION
**Pull request by k-rajat19**
Resolves https://github.com/mui/mui-x/issues/13381

This PR adds a third parameter `rowId` to `processRowUpdate` method. However, one can use `apiRef.current.getRowId` to retrieve the id of the row. Still, the more reliable and efficient approach is to pass row id as a parameter as @cherniavskii mentioned [here](https://github.com/mui/mui-x/issues/13381#issuecomment-2173104998).
By doing so, one can opt out of adding `apiRef` to the dependency array in `useCallback`, which may improve performance if the method is expensive.

- [x] I have followed (at least) the [PR section of the contributing guide](https://github.com/mui/mui-x/blob/HEAD/CONTRIBUTING.md#sending-a-pull-request).

### CODE REVIEW

File: packages/x-data-grid/src/models/props/DataGridProps.ts [hunk 1]
```diff
@@ -787,9 +787,10 @@ export interface DataGridPropsWithoutDefaultValue<R extends GridValidRowModel =
    * @template R
    * @param {R} newRow Row object with the new values.
    * @param {R} oldRow Row object with the old values.
+   * @param {GridRowId} rowId The id of the row.
    * @returns {Promise<R> | R} The final values to update the row.
    */
-  processRowUpdate?: (newRow: R, oldRow: R) => Promise<R> | R;
+  processRowUpdate?: (newRow: R, oldRow: R, rowId?: GridRowId) => Promise<R> | R;
```
Comment (cherniavskii) [inline packages/x-data-grid/src/models/props/DataGridProps.ts:793]: Why optional though? It seems that we can guarantee `rowId` presence.

Comment (k-rajat19) [reply packages/x-data-grid/src/models/props/DataGridProps.ts]: Yes we can 👍

File: packages/x-data-grid/src/models/props/DataGridProps.ts [hunk 2]
```diff
@@ -787,9 +787,10 @@ export interface DataGridPropsWithoutDefaultValue<R extends GridValidRowModel =
    * @template R
    * @param {R} newRow Row object with the new values.
    * @param {R} oldRow Row object with the old values.
+   * @param {GridRowId} rowId The id of the row.
    * @returns {Promise<R> | R} The final values to update the row.
    */
-  processRowUpdate?: (newRow: R, oldRow: R) => Promise<R> | R;
+  processRowUpdate?: (newRow: R, oldRow: R, rowId: GridRowId) => Promise<R> | R;
```
Comment (flaviendelangle) [inline packages/x-data-grid/src/models/props/DataGridProps.ts:793]: Not a huge fan of having 3 flat params with no apparent hierarchy. 
If the method is not super frequently called, maybe putting the id in a `additionalParams` object would make sure that we stick with 3 params even if we have more info to pass to this function in the future.

Comment (k-rajat19) [reply packages/x-data-grid/src/models/props/DataGridProps.ts]: thanks for the suggestion
the method is called on stop editing and copy paste, I think we should implement this change.

**Review by cherniavskii — APPROVED** [review]
Thanks!

**Pull request merged**